vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc

```diff
@@ -0,0 +1,390 @@
+      creative_instance_id,
+      [=](const bool success,
+          const absl::optional<DepositInfo>& deposit_optional) {
+        ASSERT_TRUE(success);
+
+        EXPECT_TRUE(deposit_optional);
+      });
+}
+
+void ExpectConversionCountEquals(const size_t expected_count) {
+  database::table::Conversions database_table;
+  database_table.GetAll(
+      [=](const bool success, const ConversionList& conversions) {
+        ASSERT_TRUE(success);
+
+        EXPECT_EQ(expected_count, conversions.size());
+      });
+}
+
+}  // namespace
+
+class BatAdsSearchResultAdTest : public SearchResultAdObserver,
+                                 public UnitTestBase {
+ protected:
+  BatAdsSearchResultAdTest()
+      : search_result_ad_(std::make_unique<SearchResultAd>()) {
+    search_result_ad_->AddObserver(this);
+  }
+
+  ~BatAdsSearchResultAdTest() override = default;
+
+  void OnSearchResultAdServed(const SearchResultAdInfo& ad) override {
+    ad_ = ad;
+    did_serve_ad_ = true;
+  }
+
+  void OnSearchResultAdViewed(const SearchResultAdInfo& ad) override {
+    ad_ = ad;
+    did_view_ad_ = true;
+  }
+
+  void OnSearchResultAdClicked(const SearchResultAdInfo& ad) override {
+    ad_ = ad;
+    did_click_ad_ = true;
+  }
+
+  void OnSearchResultAdEventFailed(
+      const SearchResultAdInfo& ad,
+      const mojom::SearchResultAdEventType event_type) override {
+    did_fail_to_fire_event_ = true;
+  }
+
+  std::unique_ptr<SearchResultAd> search_result_ad_;
+
+  SearchResultAdInfo ad_;
+  bool did_serve_ad_ = false;
+  bool did_view_ad_ = false;
+  bool did_click_ad_ = false;
+  bool did_fail_to_fire_event_ = false;
+};
+
+TEST_F(BatAdsSearchResultAdTest, FireViewedEvent) {
+  // Arrange
+  ForcePermissionRules();
+
+  const mojom::SearchResultAdPtr ad_mojom =
+      BuildAd(kPlacementId, kCreativeInstanceId);
+
+  // Act
+  search_result_ad_->FireEvent(ad_mojom,
+                               mojom::SearchResultAdEventType::kViewed);
+
+  // Assert
+  EXPECT_TRUE(did_serve_ad_);
+  EXPECT_TRUE(did_view_ad_);
+  EXPECT_FALSE(did_click_ad_);
+  EXPECT_FALSE(did_fail_to_fire_event_);
+  const SearchResultAdInfo& expected_ad = BuildSearchResultAd(ad_mojom);
+  EXPECT_EQ(expected_ad, ad_);
+
+  ExpectAdEventCountEquals(ConfirmationType::kViewed, 1);
+  ExpectDepositExistsForCreativeInstanceId(kCreativeInstanceId);
+  ExpectConversionCountEquals(0);
+}
+
+TEST_F(BatAdsSearchResultAdTest, FireViewedEventWithConversion) {
+  // Arrange
+  ForcePermissionRules();
+
+  const mojom::SearchResultAdPtr ad_mojom =
+      BuildAdWithConversion(kPlacementId, kCreativeInstanceId);
+
+  // Act
+  search_result_ad_->FireEvent(ad_mojom,
+                               mojom::SearchResultAdEventType::kViewed);
+
+  // Assert
+  EXPECT_TRUE(did_serve_ad_);
+  EXPECT_TRUE(did_view_ad_);
+  EXPECT_FALSE(did_click_ad_);
+  EXPECT_FALSE(did_fail_to_fire_event_);
+  const SearchResultAdInfo& expected_ad = BuildSearchResultAd(ad_mojom);
```

Comment:
nit: extra reference
and at some other places here

Reply:
Fixed